scipy/_lib/_array_api.py

```diff
@@ -195,6 +196,109 @@ def _asarray(
     return array
 
 
+def _resolve_device(array, xp_out, device=None):
+    if device is not None:
+        return device
+
+    xp_in = array_namespace(array)
+
+    # Here is what the logic is intended to be
+    # if is_numpy(xp_out) or is_array_api_strict(xp_out) or is_cupy(xp_out):
+    #     # These libraries only support one type of device
+    #     return xp_out.__array_namespace_info__().default_device()
+    # else:
+    #     return xp_in.device(array)
+
+    # but there are many issues with device support right now. For example:
+    # - NumPy `from_dlpack` only understands device='cpu' (not a device object)
+    # - CuPy `from_dlpack` doesn't work
+    # - the format of the return of `device` seems different in different libraries
+
+    # So in the meantime:
+    if is_numpy(xp_out):
+        return "cpu"
+    elif is_array_api_strict(xp_out):
+        return xp_out.__array_namespace_info__().default_device()
+    else:
+        return None
+
+
+def xp_asarray(
+        array: ArrayLike,
+        dtype: Any = None,
+        order: Literal['K', 'A', 'C', 'F'] | None = None,
+        copy: bool | None = None,
+        *,
+        xp: ModuleType | None = None,
+        device: str | None = None,
+        check_finite: bool = False,
+        subok: bool = False,
+    ) -> Array:
+    """SciPy-specific replacement for `xp.asarray` with `order`, `check_finite`,
+    `subok`, and automatic array type conversion and device transfer.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    `check_finite` is also not a keyword in the array API standard; included
+    here for convenience rather than that having to be a separate function
+    call inside SciPy functions.
+
+    `subok` is included to allow this function to preserve the behaviour of
+    `np.asanyarray` for NumPy based inputs.
+
+    """
+    xp_in = array_namespace(array)
+    if xp is None:
+        xp = xp_in
+
+    if is_numpy(xp_in):
+        # If object is array-like (but not array), make it an ndarray; otherwise, no-op.
+        array = np.asanyarray(array)
+
+    # Start with special cases that shouldn't be needed, but are due to
+    # shortcomings in library implementations of `device` and `from_dlpack`
+    if is_cupy(xp) or is_torch(xp):
+        # Torch from_dlpack doesn't work with any library - not even torch to torch
+        # CuPy `from_dlpack` doesn't accept a device argument, and when calling
+        # without one, it recommends:
+        # Use `cupy.array(numpy.from_dlpack(input))` instead.
+        array = xp.asarray(np_compat.from_dlpack(array).copy(), dtype=dtype, copy=copy)
```

Comment:
`np_compat.from_dlpack(array)` is typically read-only, and `torch` doesn't support that as input. I probably shouldn't *always* copy, but I'm not sure if we need to split hairs about respecting `copy` until the ecosystem settles down a bit.